standards/SPIFFE_Trust_Domain_and_Bundle.md

```diff
@@ -0,0 +1,290 @@
+
+The first and most obvious is that you are assuming trust in the public certificate authorities and in their fulfillment procedures. Any public CA can sign a certificate for the domain name where your bundle endpoint is hosted that will appear valid. While this may be an unacceptable risk to some, most users are OK with it given the convenience it provides and the historical stability of the public CA infrastructure.
+
+Second, more importantly, recently introduced automated fulfillment protocols like [ACME][9] make room for attacks that were not previously possible. ACME allows publicly trusted certificates to be issued without manual intervention by performing a challenge which allows a machine to assert that it is authorized by answering an HTTP request made to the DNS name of the certificate being issued. This is significant because it means that anyone that can intercept and answer the challenge is capable of receiving a valid publicly-trusted certificate. While the risk of interception across the open internet is decidedly low, risk of interception by a neighboring machine is high. Concretely, without adequate Layer 2 security controls, SPIFFE bundle endpoints using Web PKI may be subverted by malicious software with access to the same Layer 2 network as the bundle endpoint. Operators should be careful in ensuring that Layer 2 access to SPIFFE bundle endpoints protected by Web PKI is restricted.
+
+### 6.4. Reusing Cryptographic Keys Across Trust Domains
+This specification discourages sharing cryptographic keys across trust domains since the practice degrades trust domain isolation and introduces additional security challenges. When a root key is shared across multiple trust domains, it becomes critically important that authentication and authorization implementations carefully check the trust domain component of an identity and that the trust domain component be easily and habitually expressed in authorization policies.
+
+Suppose that a naïve implementation imported (ie. fully trusted) a particular root key and that the authentication system were configured to authenticate the SPIFFE identity of any SVID which chained up to the trusted root key. If the naïve implementation is not configured to only trust a specific trust domain, then any identity issued in any trust domain could be authenticated (so long as the SVID chains up to the trusted root key). Successful authentication of an external trust domain is not a problem per se since the risk can be mitigated by a correctly implemented authorization system; however, this risk could alternatively be mitigated by simply not reusing root keys across trust domains.
```

Comment:
> however, this risk could alternatively be mitigated by simply not reusing root keys across trust domains.

I don't believe that not-reusing root keys across trust domains actually mitigates the example you gave (unless I'm misunderstanding the example).

Let's say my trust domain is `spike.name` and I'm federating with your trust domain, `evan.name`. Further, you and I do not share any root keys.  If I import your bundle, but my authentication system is configured to authenticate any identity that chains up to a trusted root key, it means that you can issue SVIDs for `spiffe://spike.name/admin`, and my system will accept them.  No amount of checking at the authorization layer will save me in that case.

Reply:
What I'm attempting to highlight here are some of the perils in using a root key/cert for multiple trust domains. I'm trying to point out that it's insufficient to simply checking whether a certificate chains up to a trusted root; that the implementation must have an explicit trust domain to trust bundle mapping as part of the authentication process. (And I believe your example also highlights this peril.) I think the difference between what you're saying and what I wrote is that I'm not making an assumption that the operator has taken an additional step to federate with another mesh. In my example, the operator has not explicitly federated with another mesh, but has simply imported the root cert to which the certificates in the mesh chain up to. This situation could arise, for example, in an environment where the mesh does not use a CA dedicate to a single trust domain, but relies on some CA which mints certificates for multiple trust domains.

Reply:
Yeah, that makes sense. It seems the high level point is that authentication systems really, really must check trust domain components when validating SPIFFE IDs. Hopefully we already make that point in the SVID specs, and it's useful to make again here.  There are several failure modes of failing to do this check including your example of a single CA issuing for multiple domains (without malice) and my example of a foreign CA forging SVIDs for the local trust domain.